Implementation of new skill to apply joint torque

Ruthrash opened the pull request:
This PR implements a new skill that applys desired joint torque passed from the frankapy server. I pass a minjerkjointtrajectory generator as the trajectory generator to the skill but it is not used at all. The desired joint torque is sent from the frankapy server through TorqueControllerSensorMessage and is applied after clipping with torque limits.

Ruthrash committed:
added torque feedback controller with which we can stream direct torque commands

Ruthrash committed:
build bug fixes for direct torque controller

Ruthrash committed:
added torque feedback controller to definitions

Ruthrash committed:
added torque feedback controller to definitions

Ruthrash committed:
resolving merge conflict

Ruthrash committed:
including joint_torque_skill and added it to run_loop.cpp

Ruthrash committed:
bug fixes in build after adding new torque skill

Ruthrash committed:
added joint torque sensor message type to definitions

Commit:
working without error

Commit:
removed bug in limiting torque

Ruthrash committed:
removed unnecessary cout

mohitsharma0690 commented on franka-interface-common/include/franka-interface-common/definitions.h:
```diff
@@ -70,9 +71,10 @@ enum class FeedbackControllerType : uint8_t {
 };
 
-// Enum for Termination Handler Types
+// Enum for Termination Handler Types#end_if 
```
extra stuff at the end

mohitsharma0690 commented on franka-interface/src/feedback_controller/cartesian_impedance_feedback_controller.cpp:
please remove this file from the review.

mohitsharma0690 commented on franka-interface/src/feedback_controller_factory.cpp:
```diff
@@ -24,7 +25,11 @@ FeedbackController* FeedbackControllerFactory::getFeedbackControllerForSkill(Sha
   switch (feedback_controller_type) {
-    
+    case FeedbackControllerType::TorqueFeedbackController:
+      std::cout<<"Trying to load controller\n";
```
remove the cout

mohitsharma0690 commented on franka-interface/src/run_loop.cpp:
```diff
@@ -103,16 +104,13 @@ void run_loop::start_new_skill(BaseSkill* new_skill) {
       memory_index);
   FeedbackController *feedback_controller =
       feedback_controller_factory_.getFeedbackControllerForSkill(feedback_controller_buffer, sensor_data_manager_);
-
```
remove these line deletions.

mohitsharma0690 commented on franka-interface/src/skills/joint_torque_skill.cpp:
```diff
@@ -0,0 +1,141 @@
+//
+// Created by Ruthrash Hari on 7/9/23
+//
```
let's remove these created headers.

mohitsharma0690 commented on franka-interface/src/skills/joint_torque_skill.cpp:
```diff
@@ -0,0 +1,141 @@
+            robot_state_data->counter_ += 1;
+            robot_state_data->time_ =  period.toSec();
+            robot_state_data->has_data_ = true;
+            robot_state_data->mutex_.unlock();
```
fix indentation

mohitsharma0690 commented on franka-interface/src/skills/joint_torque_skill.cpp:
```diff
@@ -0,0 +1,141 @@
+            }
+
+            //traj_generator_->time_ = time;//////
+            //traj_generator_->dt_ = current_period_;//////
```
remove this.

mohitsharma0690 commented on franka-interface/src/skills/joint_torque_skill.cpp:
```diff
@@ -0,0 +1,141 @@
+                traj_generator_->parse_sensor_data(robot_state);
+                termination_handler_->parse_sensor_data(robot_state);
+                sensor_data_manager->getSensorBufferGroupMutex()->unlock();
+                } catch (boost::interprocess::lock_exception) {
```
fix indentation.

mohitsharma0690 commented on franka-interface/src/skills/joint_torque_skill.cpp:
```diff
@@ -0,0 +1,141 @@
+                    lock.unlock();
+                    } 
+                }catch (boost::interprocess::lock_exception) {
+        // Do nothing
```
fix these indentations.

mohitsharma0690 commented on franka-interface/src/skills/joint_torque_skill.cpp:
```diff
@@ -0,0 +1,141 @@
+                    }
+            log_counter += 1;
+            if (log_counter % 1 == 0) {
+            pose_desired = robot_state.O_T_EE_d;
```
why is this logging required here? also pose desired may not really be correct at this stage.

Ruthrash replied:
I thought this was necessary for logging robot states. I double checked O_T_EE_d and O_T_EE are the same in this stage

mohitsharma0690 commented on franka-interface/src/skills/joint_torque_skill.cpp:
```diff
@@ -0,0 +1,141 @@
+        current_joint_rotatums_[i] = -max_joint_rotatums_[i] * safety_factor;
+      }
+    }
+    current_joint_torques_[i] = previous_joint_torques_[i] + current_joint_rotatums_[i] * period;
```
isn't this logic really discontinuous, i.e. at t = 0, we have 0 torques and then just directly at the next state we have a large torque value? 
Since this code seems to be doing `q_t  = q_{t-1} + dt * (q^{desired}_{t-1} - q_{t-1}) / dt` (basically combining 25 and 34). or am I missing something here?

Ruthrash replied:
I think this is fine,
- at t = 0 we are applying 0 torque (libfranka applies torque to cancel gravity in backend) and at t = 1, we are limiting the commanded torque by limiting joint_rotatum to a fraction of the maximum in case it is greater than maximum joint_rotatum. This should limit discontinuity. 
- Please let me know if you think otherwise or if you think this could be done in a better way.

mohitsharma0690 commented on franka-interface/src/skills/joint_torque_skill.cpp:
```diff
@@ -0,0 +1,141 @@
+    SensorDataManager* sensor_data_manager = run_loop->get_sensor_data_manager();
+
+    std::cout << "Will run the control loop\n";
+    //Unneccessary for torque skill
```
can you remove these comments.

mohitsharma0690 commented on franka-interface/src/skills/joint_torque_skill.cpp:
```diff
@@ -0,0 +1,141 @@
+
+            //if (time > 0.0) {//////
+            //    traj_generator_->get_next_step(robot_state);
+            //}
```
can you please clean up this file.

mohitsharma0690 reviewed (changes requested):
Sorry for the long delay. Thanks a lot for your contribution. The code on a glance looks ok but i am not super sure if the controls are smoothly being applied (see comment). Nevertheless, if you can clean up the code and address the comments I can quickly try it on our end before merging.

Ruthrash committed:
removed unncessary change

Ruthrash committed:
removed unncessary cout

Ruthrash committed:
removed line deletions for better readability

Ruthrash committed:
fixed indentation issues

Ruthrash committed:
fixed indentation and removed comments

Ruthrash committed:
removed merge conflicts

Ruthrash committed:
removed unnecessary changes

Ruthrash commented:
Hey @mohitsharma0690! Apologies for getting so late to reply, I was travelling for the most part of the time after your change requests and I didn't have access to the franka robot. Thanks so much for your comments, I have incoporated the changes you have asked and have addressed some of your concerns. Please get back to me when you could.

Ruthrash committed:
Update make_catkin.sh to use local installation of libfranka
